### DISCUSSION TITLE
jax.value_and_grad() and CUDA kernels

### DISCUSSION THREAD

**Discussion by mah-asd** [jax-ml/jax#12450] [Q&A] [answered]
Hello, 
My question is about using jax.value_and_grad(fun) while we have a Cuda kernel call in our fun. 
Is there any way to calculate the grad in this situation?

**Comment by jakevdp** [comment, marked as answer]
The way to support custom CUDA calls would be via a `custom_jvp`, which you can read about here: https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html

Essentially, you need to define a function that computes the gradient for your kernel call, and register it so that the JAX autodiff machinery knows to use that function when it encounters your Cuda kernel during tracing. Hopefully that will point you in the right direction!

**Reply by mah-asd** [reply]
Thank you so much!
